### PR TITLE
[lldb] Improved lldb-server stability for remote launching

### PR DESCRIPTION
**Pull request by slydiman**
We faced the issue running cross api tests in 8 threads. The executable is installed to the target by the process `lldb-server platform`, but launched by the another process `lldb-server gdbserver`. We got the error ETXTBSY on Linux and ERROR_SHARING_VIOLATION on Windows. It seems the known issue and ProcessLauncherPosixFork.cpp already contains the workaround, but it is not enough. Updated the workaround with the total timeout 5 seconds and added the same workaround to ProcessLauncherWindows.cpp too.

### CODE REVIEW

File: lldb/source/Host/posix/ProcessLauncherPosixFork.cpp [hunk 1]
```diff
@@ -201,7 +201,7 @@ struct ForkLaunchInfo {
   execve(info.argv[0], const_cast<char *const *>(info.argv), info.envp);
 
 #if defined(__linux__)
-  if (errno == ETXTBSY) {
+  for (int i = 0; i < 50; ++i) {
```
Comment (bulbazord) [inline lldb/source/Host/posix/ProcessLauncherPosixFork.cpp:204]: Maybe it would be a good idea to make the timeout configurable?

Comment (slydiman) [reply lldb/source/Host/posix/ProcessLauncherPosixFork.cpp:204]: I think it is redundant. Currently lldb-server contains a hardcoded timeout in many places. Note the total timeout 10 seconds will cause the error "Sending vRun packet failed". So 5 seconds is good enough for the stable connection and do the trick on slow machines.

**Review by labath — CHANGES_REQUESTED** [review]
The workaround specifically mentions Android (M). The problem there was that the ADB daemon (part of the OS, which we use for faster file uploads) was implemented (had a bug) such that the file could remain open for a short while (between fork() and exec()) in another process when the daemon is accepting a new connection.

We fixed that issue, but we since the this comes with the device, we still had to work around the problem in old devices. We're finally reaching the point when we could remove the workaround (android <=M is less that 3% of active devices), so I wouldn't want to extend it without understanding the problem further.

As I understand it, you're using the regular `lldb-server platform` process to implement file uploads, which I think means the problem should be fully within our control. We basically need to make sure that:
- lldb waits for the `vFile:close` response before sending the `vRun` packet
- lldb-server actually closes the file handle (all of them -- i.e., makes sure it does not leak it) before sending the `vFile:close` response

If that's not what's happening right now, then we ought to fix it.

**Comment by labath** [comment]
> If that's not what's happening right now, then we ought to fix it.

Or at least, understand why is that impossible.

**Comment by slydiman** [comment]
Of course lldb waits for the vFile:close response before sending the vRun packet and lldb-server actually closes the file handle (all of them). No any leaks. Otherwise this workaround wouldn't work. 

The behavior is the same on Linux and Windows targets. I launched 100 connections and 200 processes simultaneously on Windows (lldb-server gdbserver + a test app). I got 3..10 fails because of  the error ERROR_SHARING_VIOLATION. After this patch I got 0..3 fails for 100 connections and 0 fails for 50 connections. After closing the copied file probably the system may cache it some time per process. The file may be blocked by the built-in antivirus for some time. It is hard to figure out the exact reason.

We have a buildbot to run cross API tests in 8 threads with Linux Aarch64 target. All tests are green with the current (single thread) lldb-server. But we got randomly failed 50..60 tests with the multithreading version of lldb-server. Probably it is just little bit faster and system did not unlock the executable. We noticed that usually failed tests use simple and tiny executables. But this fact does not help to explain the reason of the problem. We got 100% green tests after this patch.

**Comment by labath** [comment]
That *may* be how things works on windows(*), but I'm pretty sure it's not how things work on linux. A much more likely scenario is:
1. thread 1 opens a file for writing (file descriptor A) and starts writing it
2. thread 2 starts launching a gdb server. It calls fork(), which creates another process with a copy of fd A (the fd has the CLOEXEC flag set, but not CLOFORK (the flag doesn't exist))
3. thread 1 finishes writing, closes fd A (in its/parent process)
4. thread 1launches gdb-server, gdb-server tries to launch the file it has just written, gets ETXTBSY because the fd is still open in the child process

This isn't "operating system keeping the file open longer", this is us doing it (to be fair, the operating system is making it pretty hard to avoid doing that). And while this isn't an absolute thing (the workaround isn't as bad as the `select` thing, and it may also be possible to implement this in a way that the workaround isn't needed), I think this is a good reason to prefer a multiprocess implementation (where this situation does not occur because the fd cannot leak into another process).

> All tests are green with the current (single thread) lldb-server.

Does this refer the the forking implementation you get with the `--server` flag, or the serial implementation which only handles one connection at a time (without the flag)? Because if it's the former, this is a good proof that the scenario above is the (only) problem here.

(*) I just don't know enough about the system to have an informed opinion. This isn't the first time I've heard about the antivirus hypothesis, but I find that somewhat surprising, as that could mean that something like `$CC hello.cc && ./a.out` could fail (what we're doing here isn't fundamentally different than that).

**Comment by labath** [comment]
> * thread 1 opens a file for writing (file descriptor A) and starts writing it
> 
> * thread 2 starts launching a gdb server. It calls fork(), which creates another process with a copy of fd A (the fd has the CLOEXEC flag set, but not CLOFORK (the flag doesn't exist))
> 
> * thread 1 finishes writing, closes fd A (in its/parent process)
> 
> * thread 1launches gdb-server, gdb-server tries to launch the file it has just written, gets ETXTBSY because the fd is still open in the child process

In this scenario, waiting does make the exec succeed, because the FD in the forked process will not stay open very long. It will get closed as soon as the process runs `execve` (due to CLOEXEC).

**Comment by slydiman** [comment]
> Does this refer the the forking implementation you get with the --server flag, or the serial implementation which only handles one connection at a time (without the flag)?

I mean the `--server` flag. It is very hard to reproduce this issue with the serial implementation because it is much slower.

> - thread 1 opens a file for writing (file descriptor A) and starts writing it
> - thread 2 starts launching a gdb server. It calls fork(), which creates another process with a copy of fd A (the fd has the CLOEXEC flag set, but not CLOFORK (the flag doesn't exist))
> - thread 1 finishes writing, closes fd A (in its/parent process)
> - thread 1launches gdb-server, gdb-server tries to launch the file it has just written, gets ETXTBSY because the fd is still open in the child process

> In this scenario, waiting does make the exec succeed, because the FD in the forked process will not stay open very long. It will get closed as soon as the process runs execve (due to CLOEXEC).

Your scenario is impossible. How FD will be closed by execve() if execve() failed with ETXTBSY?

- `lldb-server platform` (a `thread 1` or just a process) creates, writes and closes the file. The FD is closed and may be reused by the system. The client lldb received the response OK for `vFile:close` request.
- The same `thread 1` launched `gdb server` (fork+execve). The FD of a created and closed file cannot be copied any way.
- The client lldb connects to `gdb server` and send the request `vRun`
- `gdb server` did not create or write the executable. It never had the FD of this file and did not inherit it. `gdb server` just fork a child process and try to call execve(), but it failed with ETXTBSY.

> `$CC hello.cc && ./a.out` could fail

May it work? How is it possible?

**Comment by slydiman** [comment]
> that could mean that something like `$CC hello.cc && ./a.out` could fail (what we're doing here isn't fundamentally different than that).

The difference is that cc creates a.out and exits itself. But `lldb-server platform` is still running after creating the executable. Something must be flushed?

**Comment by slydiman** [comment]
See also https://github.com/golang/go/issues/22315

**Comment by labath** [comment]
> > Does this refer the the forking implementation you get with the --server flag, or the serial implementation which only handles one connection at a time (without the flag)?
> 
> I mean the `--server` flag. It is very hard to reproduce this issue with the serial implementation because it is much slower.

Ok, so if I'm reading this right you're saying you saw no ETXTBSY errors with the current implementation `--server` flag. Is that correct ?

> > * thread 1 opens a file for writing (file descriptor A) and starts writing it
> > * thread 2 starts launching a gdb server. It calls fork(), which creates another process with a copy of fd A (the fd has the CLOEXEC flag set, but not CLOFORK (the flag doesn't exist))
> > * thread 1 finishes writing, closes fd A (in its/parent process)
> > * thread 1launches gdb-server, gdb-server tries to launch the file it has just written, gets ETXTBSY because the fd is still open in the child process
> 
> > In this scenario, waiting does make the exec succeed, because the FD in the forked process will not stay open very long. It will get closed as soon as the process runs execve (due to CLOEXEC).

There are two threads and (at least two execve()s) happening here. I'm referring to the one on thread 2 (specifically, the fork child of thread 2).  Your description describes what happens on one thread (well, one line of execution, corresponding to one e.g.  test). Let me try this again. I'm just going to take your description, copy it twice and interleave it (*italic* is for one line of execution **bold** is for the second one, regular text is my commentary):




1. *`lldb-server platform` (a `thread 1` or just a process) creates, writes* 
2. note that at this point the file remains open in the lldb-platform process. This is going to be our mythical FD
3. **`lldb-server platform` (a `thread 1` or just a process) creates, writes and closes the file. The FD is closed and may be reused by the system. The client lldb received the response OK for `vFile:close` request.**
4.  **The same `thread 1` launched `gdb server` (fork**
5. note the fork creates a new process. The process is going to have a copy of the mythical FD
6. *and closes the file. The FD is closed and may be reused by the system. The client lldb received the response OK for `vFile:close` request.*
7. note the mythical FD is only closed in the parent process. It still exists in the fork child
8. *The same `thread 1` launched `gdb server` (fork+execve). The FD of the created and closed file cannot be copied any way.*
9. We don't need to make a copy of the fd here. The copy was made earlier.
10. *`gdb server` did not create or write the executable. It never had the FD of this file and did not inherit it. `gdb server` just fork a child process and try to call execve(), but it failed with ETXTBSY.*
11. We get ETXTBSY because the FD is still open in the process forked on step 4
12. **+execve). The FD of the created and closed file cannot be copied any way.**
13. The mythical fd is fully closed after the execve() call
14. *`gdb server` waits little bit and try to call execve() again. It is successful after several attempts. No one closed the mythical FD during this time.*
15. Yes, they did.

**Comment by slydiman** [comment]
> Ok, so if I'm reading this right you're saying you saw no ETXTBSY errors with the current implementation --server flag. Is that correct ?

Right. Initially I have marked #100670 as dependent on this.

Ok, agreed. So, we can try to use O_CLOFORK. 
And simple solution is to call execve() as fast as possible and wait some time in case of ETXTBSY (this PR).

**Comment by labath** [comment]
> > Ok, so if I'm reading this right you're saying you saw no ETXTBSY errors with the current implementation --server flag. Is that correct ?
> 
> Right. Initially I have marked #100670 as dependent on this.
> 
> Ok, agreed. So, we can try to use O_CLOFORK. And simple solution is to call execve() as fast as possible and wait some time in case of ETXTBSY (this PR).

Umm.. by "current" I meant the current implementation that's in the llvm repository, so I'm not sure if we're agreeing to anything (yet).

O_CLOFORK doesn't exist (that's the really mythical part). I wish it did though...

I don't think we can call execve appreciably faster than we already do. I'm still not sure if I am ok with the wait workaround, but I think it could wait until we settle some other things first. For one, I'd like to hear your opinion on my port mapping alternative.

**Comment by slydiman** [comment]
> I'd like to hear your opinion on my port mapping alternative.

https://github.com/llvm/llvm-project/pull/100670#issuecomment-2255991921

**Comment by slydiman** [comment]
I have moved this patch to #100670.

**Pull request closed without merging**